### PR TITLE
NIFI-9951 Add proxy support to Azure ADLS and Blob v12 processors

### PR DESCRIPTION
**Pull request by nandorsoma**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Add proxy support to Azure ADLS and Blob v12 processors

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by kulikg** [comment]
Reviewing

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java [hunk 1]
```diff
@@ -311,4 +316,38 @@ public static void setProxy(final OperationContext operationContext, final Proce
...
+
+            nettyClientBuilder.proxy(proxyOptions);
+        }
+    }
+
+    private static ProxyOptions.Type getProxyType(ProxyConfiguration proxyConfiguration) {
+        if (proxyConfiguration.getProxyType() == Proxy.Type.HTTP) {
+            return ProxyOptions.Type.HTTP;
+        } else if (proxyConfiguration.getProxyType() == Proxy.Type.SOCKS) {
+            SocksVersion socksVersion = proxyConfiguration.getSocksVersion();
```
Comment (kulikg) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java:343]: Please consider to make this variable final.

File: nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-bundle/nifi-proxy-configuration/src/main/java/org/apache/nifi/proxy/StandardProxyConfigurationService.java [hunk 2]
```diff
@@ -95,7 +106,10 @@ protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
     @OnEnabled
     public void setConfiguredValues(final ConfigurationContext context) {
         configuration = new ProxyConfiguration();
-        configuration.setProxyType(Proxy.Type.valueOf(context.getProperty(PROXY_TYPE).getValue()));
+
+        Proxy.Type proxyType = Proxy.Type.valueOf(context.getProperty(PROXY_TYPE).getValue());
```
Comment (kulikg) [inline nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-bundle/nifi-proxy-configuration/src/main/java/org/apache/nifi/proxy/StandardProxyConfigurationService.java:110]: Please consider to make this variable final.

**Review by kulikg — COMMENTED** [review]
Please take a look into my suggestions.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java [hunk 3]
```diff
@@ -311,4 +316,38 @@ public static void setProxy(final OperationContext operationContext, final Proce
...
+
+    private static ProxyOptions.Type getProxyType(ProxyConfiguration proxyConfiguration) {
+        if (proxyConfiguration.getProxyType() == Proxy.Type.HTTP) {
+            return ProxyOptions.Type.HTTP;
+        } else if (proxyConfiguration.getProxyType() == Proxy.Type.SOCKS) {
+            final SocksVersion socksVersion = proxyConfiguration.getSocksVersion();
+            if (socksVersion != SocksVersion.NOT_SPECIFIED) {
+                return ProxyOptions.Type.valueOf(socksVersion.name());
+            } else {
+                throw new IllegalArgumentException("Version should be set when SOCKS proxy type is used!");
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java:347]: Use of exclamations in error messages should be avoided. Recommend adjusting the wording for greater clarity:
```suggestion
                throw new IllegalArgumentException("SOCKS Version is required for SOCKS Proxy Type");
```

File: nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-bundle/nifi-proxy-configuration/src/main/java/org/apache/nifi/proxy/StandardProxyConfigurationService.java [hunk 4]
```diff
@@ -46,31 +46,41 @@ public class StandardProxyConfigurationService extends AbstractControllerService
             .required(true)
             .build();
 
-    static final PropertyDescriptor PROXY_SERVER_HOST = new PropertyDescriptor.Builder()
+    public static final PropertyDescriptor SOCKS_VERSION = new PropertyDescriptor.Builder()
+        .name("socks-version")
+        .displayName("Socks version")
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-bundle/nifi-proxy-configuration/src/main/java/org/apache/nifi/proxy/StandardProxyConfigurationService.java:51]: `SOCKS` should be uppercase:
```suggestion
        .displayName("SOCKS Version")
```

File: nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-bundle/nifi-proxy-configuration/src/main/java/org/apache/nifi/proxy/StandardProxyConfigurationService.java [hunk 5]
```diff
@@ -46,31 +46,41 @@ public class StandardProxyConfigurationService extends AbstractControllerService
...
+        .description("Socks version.")
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-bundle/nifi-proxy-configuration/src/main/java/org/apache/nifi/proxy/StandardProxyConfigurationService.java:52]: Recommend expanding the description:
```suggestion
        .description("SOCKS Protocol Version")
```

File: nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-api/src/main/java/org/apache/nifi/proxy/SocksVersion.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.proxy;
+
+public enum SocksVersion {
+    NOT_SPECIFIED,
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-api/src/main/java/org/apache/nifi/proxy/SocksVersion.java:20]: Recommend removing `NOT_SPECIFIED` as a value on this enum and defaulting to `SOCKS5` since it should be more commonly used.

File: nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-bundle/nifi-proxy-configuration/src/main/java/org/apache/nifi/proxy/StandardProxyConfigurationService.java [hunk 7]
```diff
@@ -46,31 +46,41 @@ public class StandardProxyConfigurationService extends AbstractControllerService
...
+        .allowableValues(SocksVersion.values())
+        .defaultValue(SocksVersion.NOT_SPECIFIED.name())
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-bundle/nifi-proxy-configuration/src/main/java/org/apache/nifi/proxy/StandardProxyConfigurationService.java:54]: Recommend changing the default value to `SOCKS5`.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for implementing this improvement @nandorsoma! I noted a few minor wording adjustments for some messages and properties.

The addition of the `SOCKS Version` property seems reasonable, given that the Azure Netty client requires the version to be specified. The Java Proxy.Type is not specific, and the implementation attempts to negotiate the version, preferring SOCKS5. It seems better to default to SOCKS5, as opposed to unspecified, since that should be commonly available for SOCKS proxy servers.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureBlobProcessor_v12.java [hunk 8]
```diff
@@ -83,45 +87,38 @@ public abstract class AbstractAzureBlobProcessor_v12 extends AbstractProcessor {
             .description("Unsuccessful operations will be transferred to the failure relationship.")
             .build();
 
+    protected static final List<PropertyDescriptor> PROPERTIES = Collections.unmodifiableList(Arrays.asList(
+            STORAGE_CREDENTIALS_SERVICE,
+            AzureStorageUtils.PROXY_CONFIGURATION_SERVICE
+    ));
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureBlobProcessor_v12.java:93]: The abstract processor did not contain common properties intentionally. I strongly recommend to use simple property lists in processor classes instead. This way each processor can define its properties in the right order and without cumbersome code (adding/deleting properties).

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureBlobProcessor_v12.java [hunk 9]
```diff
@@ -83,45 +87,38 @@ public abstract class AbstractAzureBlobProcessor_v12 extends AbstractProcessor {
...
+        clientBuilder.httpClient(nettyClient);
+
         configureCredential(clientBuilder, credentialsService, credentialsDetails);
 
         return clientBuilder.buildClient();
     }
 
-    private static void configureCredential(BlobServiceClientBuilder clientBuilder, AzureStorageCredentialsService_v12 credentialsService, AzureStorageCredentialsDetails_v12 credentialsDetails) {
+    private static void configureCredential(BlobServiceClientBuilder clientBuilder, AzureStorageCredentialsService_v12 credentialsService,
+                                            AzureStorageCredentialsDetails_v12 credentialsDetails) {
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureBlobProcessor_v12.java:121]: Max line length is 200. Please do not apply unnecessary reformatting.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureBlobProcessor_v12.java [hunk 10]
```diff
@@ -131,7 +128,7 @@ private static void configureCredential(BlobServiceClientBuilder clientBuilder,
                 break;
             case MANAGED_IDENTITY:
                 clientBuilder.credential(new ManagedIdentityCredentialBuilder()
-                                .clientId(credentialsDetails.getManagedIdentityClientId())
+                        .clientId(credentialsDetails.getManagedIdentityClientId())
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/AbstractAzureBlobProcessor_v12.java:131]: This is a good example for fixing formatting. Thanks!

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java [hunk 11]
```diff
@@ -83,18 +83,36 @@ public class PutAzureDataLakeStorage extends AbstractAzureDataLakeStorageProcess
...
+    ));
 
-    @Override
-    protected void init(final ProcessorInitializationContext context) {
-        final List<PropertyDescriptor> props = new ArrayList<>(super.getSupportedPropertyDescriptors());
-        props.add(CONFLICT_RESOLUTION);
-        properties = Collections.unmodifiableList(props);
+    public static long MAX_CHUNK_SIZE = 100 * 1024 * 1024; // current chunk limit is 100 MiB on Azure
+
+    static void uploadContent(DataLakeFileClient fileClient, InputStream in, long length) {
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java:92]: Please do not reorder class members because it makes the git history hard to follow.

**Review by turcsanyip — CHANGES_REQUESTED** [review]
@nandorsoma Thanks for adding proxy capability to these processors!

I checked the processors' properties and found some inconsistencies:
- `ListAzureDataLakeStorage` does not have `Proxy Configuration Service` property.
- `Proxy Configuration Service` property appears in different positions in the processors. In my opinion it should be the last one as in case of ListAzureBlobStorage_v12.

**Comment by nandorsoma** [comment]
@exceptionfactory @turcsanyip Thank you for your review. I've changed what you've asked.

File: nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-api/src/main/java/org/apache/nifi/proxy/SocksVersion.java (hunk 6 above)
Comment (nandorsoma) [reply nifi-nar-bundles/nifi-standard-services/nifi-proxy-configuration-api/src/main/java/org/apache/nifi/proxy/SocksVersion.java]: TLDR
The original problem is that the actual proxy client implementation used in the processor is loosely coupled from the proxy configuration service. That can lead to problems, for example you can choose a `Proxy.Type` in the service which won't be supported in the processor. I think this is a known problem, that's why as a workaround you can supply the supported `Proxy.Type` to `ProxyConfiguration.createProxyConfigPropertyDescriptor()` which will be used to generate - unfortunately only - a description. Because of that, in my very first solution I wanted to create a new configuration service for such processors that supports `SOCKS4` and `SOCKS5`. Meanwhile I learned that it is a rule of thumb to have only one proxy configuration service in a flow so I've discarded that solution. Though I still felt that it would be misleading to specify `SocksVersion` for a processor where the property will be ignored. That's why I wanted to explicitly have `NOT_SPECIFIED` to somehow represent this intention.

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java [hunk 12]
```diff
@@ -311,4 +316,34 @@ public static void setProxy(final OperationContext operationContext, final Proce
         final ProxyConfiguration proxyConfig = ProxyConfiguration.getConfiguration(processContext);
         operationContext.setProxy(proxyConfig.createProxy());
     }
+
+    public static void configureProxy(final NettyAsyncHttpClientBuilder nettyClientBuilder, final PropertyContext propertyContext) {
```
[the remark this answers is not in the document]
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java:320]: That's a good point @turcsanyip, I agree that it would be better to decouple this utility method from the Netty Client implementation.  This PR does couple the components to the Netty implementation, where previously the implementation was loaded based on available libraries. Not sure of other impacts, but something to consider.

[the remark this answers is not in the document]
Comment (nandorsoma) [reply nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/AzureStorageUtils.java:320]: Hey! I just wanted to follow a pattern that I've seen in the class. (`AbstractAzureBlobProcessor_v12.configureCredential()`) But apart from that I agree with you. Changed it!

**Comment by nandorsoma** [comment]
Hey @turcsanyip! I've changed what you asked here and I've also added the property refactor that we were talking about in pm.

**Review by turcsanyip — APPROVED** [review]
@nandorsoma Thanks for the review changes!
The code looks good to me now. I also tested the processors both with HTTP and SOCKS proxies. All work as expected.

+1 LGTM

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for making the adjustments @nandorsoma! The latest version looks good. If you can resolve the conflicts, it should be ready to merge.

**Review by exceptionfactory — APPROVED** [review]
+1 pending resolution of conflicts.

**Comment by nandorsoma** [comment]
Thank you @turcsanyip and @exceptionfactory! I've rebased the branch and quickly run a sanity test with PutAzureBlobStorage_v12 and ListAzureBlobStorage_v12 HTTP proxy enabled.

**Comment by turcsanyip** [comment]
@nandorsoma Thanks for the review changes and the rebase!
@exceptionfactory @kulikg Thanks for your reviews!

Merging to main.
